src/ThriveWorkerUnit.sol

```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {EnumerableSet} from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "./IBadgeQuery.sol";
+
+/**
+ * @title ThriveWorkerUnit
+ * @dev Contract of a work unit representing a task where contributors complete and earn rewards.
+ */
+contract ThriveWorkerUnit {
+    using SafeERC20 for IERC20;
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using EnumerableSet for EnumerableSet.Bytes32Set;
+
+    address public immutable moderator;
+    address public immutable rewardToken;
+    uint256 public immutable rewardAmount;
+    uint256 public immutable maxRewards;
+    uint256 public immutable validationRewardAmount;
+    address public assignedAddress;
+    uint256 public maxCompletionsPerUser;
+    uint256 public deadline;
+    string public validationMetadata;
+    string public metadataVersion;
+    string public metadata;
+
+    EnumerableSet.AddressSet private validators;
+    EnumerableSet.Bytes32Set private requiredBadges;
+
+    mapping(address => uint256) public completions;
+
+    IBadgeQuery public badgeQuery;
+    bool public ready;
+
+    event Initialized();
+    event ConfirmationAdded(
+        address indexed contributor,
+        string validationMetadata,
+        uint256 rewardAmount,
+        address indexed validator
+    );
+    event Withdrawn(address token, uint256 amount);
+
+    modifier onlyModerator() {
+        require(msg.sender == moderator, "ThriveProtocol: only the moderator can perform this action");
+        _;
+    }
+
+    modifier onlyValidator() {
+        require(validators.contains(msg.sender), "ThriveProtocol: only a validator can perform this action");
+        _;
+    }
+
+    modifier onceReady() {
+        require(ready, "ThriveProtocol: contract is not ready");
+        _;
+    }
+
+    constructor(
+        address _moderator,
+        address _rewardToken,
+        uint256 _rewardAmount,
+        uint256 _maxRewards,
+        uint256 _deadline,
+        uint256 _maxCompletionsPerUser,
+        address[] memory _validators,
+        string memory _validationMetadata,
+        string memory _metadataVersion,
+        string memory _metadata,
+        address _badgeQuery
+    ) {
+        require(_moderator != address(0), "ThriveProtocol: moderator address is required");
+        require(_badgeQuery != address(0), "ThriveProtocol: badgeQuery address is required");
+        require(_deadline > block.timestamp, "ThriveProtocol: deadline must be in the future");
+
+        moderator = _moderator;
+        rewardToken = _rewardToken;
+        rewardAmount = _rewardAmount;
+        maxRewards = _maxRewards;
+        validationRewardAmount = _rewardAmount / 10; // Validator Reward = 10% of contributor reward
```

Comment:
hmm this won't work, reward amount is in erc20 but validator reward amount is in thrive and I think this should be flexible and determined by creator

Reply:
should we go with a setter of validationRewardAmount? or just the percentage?